src/test/python/test_signer.py

```diff
@@ -0,0 +1,183 @@
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import pytest
+from covata.delta import signer
+import json
+import re
+from datetime import datetime
+import six.moves.urllib as urllib
+from freezegun import freeze_time
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric import padding
+
+freeze_time("2017-01-03 14:48:10")
+
+SIGNING_ALGORITHM = "CVT1-RSA4096-SHA256"
+CVT_DATE_FORMAT = "%Y%m%dT%H%M%SZ"
+
+
+@pytest.mark.parametrize('payload, expected_hash', [
+    (b'{"name": "user"}',
+     '5077ee49430b0c34347573ca6d189b29fc98cf15b63b74f82460cf46ac1bb0a5'),
+    (b'{}', '44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a'),
+])
+def test_get_hashed_payload(payload, expected_hash):
+    assert signer.__get_hashed_payload(payload) == expected_hash
+
+
+def test_malformed_payload():
+    with pytest.raises(ValueError) as excepinfo:
+        signer.__get_hashed_payload(b'{""}')
+        assert excepinfo.type is json.JSONDecodeError
+
+
+def test_none_payload():
+    assert signer.__get_hashed_payload(b'{}') == \
+           signer.__get_hashed_payload(None)
+
+
+def test_unordered_payload():
+    assert signer.__get_hashed_payload(b'{"name": "rattan", "a": "value"}') == \
+           signer.__get_hashed_payload(b'{"a": "value", "name": "rattan"}')
+
+
+def test_nested_payload():
+    input_json = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    sorted_json = \
+        b"""
+        {
+            "a": "value",
+            "c": {
+                "a": "hello",
+                "za": "value"
+            },
+            "name": "rattan"
+        }
+        """
+
+    assert signer.__get_hashed_payload(input_json) == \
+           signer.__get_hashed_payload(sorted_json)
+
+
+@pytest.mark.parametrize('url, expected_encoded_url', [
+    ('/identity blah/', '/identity%20blah/'),
+    ('/identity-blah/', '/identity-blah/'),
+    ('/', '/'),
+    ('/identityABC/', '/identityABC/'),
+])
+def test_uri_endoing(url, expected_encoded_url):
+    assert signer.__encode_uri(url) == expected_encoded_url
+
+
+def test_signature_material(private_key):
+    method = "POST"
+    url = "https://delta.covata.io/v1/secrets?hello=w+orld"
+    headers = {
+        "Content-Type": "application/json"
+    }
+    payload = \
+        b"""
+        {
+            "name": "rattan",
+            "a": "value",
+            "c": {
+                "za": "value",
+                "a": "hello"
+            }
+        }
+        """
+
+    with freeze_time():
```

Comment:
pretty sure with pytest you can use `freeze_time` as decorator, so to get rid of the `with` and set the context to the test function instead, i.e.

```python
@freeze_time()
def test_signature_material(private_key):
```